### PR TITLE
Migrate LESS to SCSS

### PR DESCRIPTION
**Pull request by janfaracik**
This PR migrates our codebase from using LESS to SCSS.

The justification for this is as follows:

1. SCSS has a larger feature set and its syntax is (personally) more sensible
2. SCSS has a far larger community, meaning that more people will have the skillset necessary to contribute to Jenkins
3. A number of plugins already use SCSS, so this isn't unknown territory for our existing developers

https://www.telerik.com/blogs/why-bootstrap-4-means-sass-has-won

https://sproutsocial.com/insights/less-to-sass/

---

All LESS files have been converted to SCSS with the exception of pluginSetupWizard.less, which is reliant on Bootstrap 3 (only available in LESS, however newer versions are only available in SCSS). It's my intention to convert this file after this PR to remove those dependencies and to use SCSS.

### Testing done

* Jenkins should work identically to before, there should be no visual or functional difference

There will no doubt be things I have missed, so please let me know if you see any visual discrepancies. 

### Proposed changelog entries

- N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7850"></a>

### CODE REVIEW

**Comment by NotMyFault** [comment]
Hm, I've seen less usage of it.

**Comment by timja** [comment]
+1 on concept, less is dead afaik, I've only seen sass or plain CSS these days.

**Comment by NotMyFault** [comment]
Loading pages yields `localhost:8080/static/images/spinner.gif 404 (Not Found)` occasionally in the browser console.

**Comment by janfaracik** [comment]
> Loading pages yields `localhost:8080/static/images/spinner.gif 404 (Not Found)` occasionally in the browser console.

Great spot 👍

**Comment by NotMyFault** [comment]
> > Loading pages yields `localhost:8080/static/images/spinner.gif 404 (Not Found)` occasionally in the browser console.
> 
> Great spot 👍

I'll take a more detailed look next week 👏

File: war/src/main/scss/modules/notifications.scss
Comment (basil) [inline war/src/main/scss/modules/notifications.scss:6]: This produces a difference in the compiled CSS: from `max-width: min(600px, calc(100vw - 2.4rem));` to `max-width: min(600px, 100vw - 2.4rem);`. Is that intentional? If not, please disregard this comment.

File: war/src/main/scss/modules/side-panel-tasks.scss
Comment (basil) [inline war/src/main/scss/modules/side-panel-tasks.scss:53]: This adds `position: relative;` and `z-index: 0;` that were not present in the original CSS. Was this intentional?

File: war/src/main/scss/modules/buttons-deprecated.scss
Comment (basil) [inline war/src/main/scss/modules/buttons-deprecated.scss:3]: Something seems to be going wrong in this file. The following styles are missing from the generated CSS in the SCSS version, while they are present in the Less version:

```css
.button-link {
  color: var(--primary);
  background-color: transparent;
  border-color: transparent;
  font-weight: var(--btn-link-font-weight);
}
.button-link:hover,
.button-link:focus {
  color: var(--btn-link-color--hover);
  background-color: var(--btn-link-bg--hover);
  border-color: var(--btn-link-bg--hover);
}
.button-link:focus {
  box-shadow: none;
}
.button-link:active {
  color: var(--btn-link-color--active);
  background-color: var(--btn-link-bg--active);
  border-color: var(--btn-link-bg--active);
}
.button-disabled {
  opacity: 0.5;
  pointer-events: none;
}
```

File: war/package.json
Comment (basil) [inline war/package.json:51]: Is this change needed for this PR or can this hunk be reverted and left for Dependabot to minimize the scope of this change?

File: war/package.json
Comment (basil) [inline war/package.json:46]: Can we remove the carat to be consistent with the other version numbers in this file?

File: war/src/main/scss/abstracts/colors.scss
Comment (basil) [inline war/src/main/scss/abstracts/colors.scss:6]: Something seems to be going wrong here. The generated CSS file in your new version has `.jenkins-\!-color-light-key` rather than e.g. `.jenkins-\!-color-light-brown` as expected.

File: war/src/main/scss/base/spacing.scss
Comment (basil) [inline war/src/main/scss/base/spacing.scss:6]: Something seems to be going wrong here. The generated CSS file in your version has `.jenkins-\!-margin-top-value` rather than e.g. `.jenkins-\!-margin-left-2` as expected.

**Review by basil — CHANGES_REQUESTED** [review]
This PR breaks the plugin setup wizard, which references a `pluginSetupWizard.css` file which is no longer delivered in the WAR as of this PR at its latest revision.

File: war/src/main/scss/abstracts/theme.scss
Comment (janfaracik) [reply war/src/main/scss/abstracts/theme.scss]: Good spot, I've restored 0.5

File: war/src/main/scss/abstracts/colors.scss
Comment (janfaracik) [reply war/src/main/scss/abstracts/colors.scss]: Great spot, fixed.

File: war/src/main/scss/base/spacing.scss
Comment (janfaracik) [reply war/src/main/scss/base/spacing.scss:6]: Fixed :)

File: war/src/main/scss/base/style.scss
Comment (janfaracik) [reply war/src/main/scss/base/style.scss:118]: I think it's fine, the class will still exist on the page it just won't have an empty selector in the stylesheet. I've removed it from `style.scss` as there isn't any benefit in having it there.

File: war/src/main/scss/modules/notifications.scss
Comment (janfaracik) [reply war/src/main/scss/modules/notifications.scss]: Fixed :)

File: war/src/main/scss/modules/buttons.scss
Comment (janfaracik) [reply war/src/main/scss/modules/buttons.scss:11]: It appears for me?

image

File: war/src/main/scss/modules/buttons-deprecated.scss
Comment (janfaracik) [reply war/src/main/scss/modules/buttons-deprecated.scss:3]: Good spot, restored.

File: war/src/main/scss/modules/side-panel-tasks.scss
Comment (janfaracik) [reply war/src/main/scss/modules/side-panel-tasks.scss:53]: Hm they should have been there originally. I've removed the additions of them from .task-link as they were duplicating what was in mixins.item;

File: war/src/main/scss/modules/buttons.scss
Comment (basil) [reply war/src/main/scss/modules/buttons.scss:11]: The fact that it appears in the compiled CSS is the potential problem I was pointing out, since it does not appear in the compiled CSS before this PR.

File: war/src/main/scss/base/spacing.scss
Comment (basil) [inline war/src/main/scss/base/spacing.scss:29]: Why doesn't this include 0? The following styles from the old compiled CSS are now gone:

```css
.jenkins-\!-padding-0 {
  padding: 0rem !important;
}
.jenkins-\!-padding-top-0 {
  padding-top: 0rem !important;
}
.jenkins-\!-padding-left-0 {
  padding-left: 0rem !important;
}
.jenkins-\!-padding-bottom-0 {
  padding-bottom: 0rem !important;
}
.jenkins-\!-padding-right-0 {
  padding-right: 0rem !important;
}
```

If this was intentional, then please disregard this comment.

File: war/src/main/scss/base/style.scss
Comment (basil) [reply war/src/main/scss/base/style.scss:118]: Sounds good, intentional changes like these make sense to me.

File: war/src/main/scss/abstracts/theme.scss
Comment (basil) [inline war/src/main/scss/abstracts/theme.scss:25]: The `.app-theme-picker__picker` CSS used to be present in the compiled CSS for `add-page.css`, `simple-page.css`, and `styles.css`. If its removal from `add-page.css` was intentional then please ignore this comment. If it needs to remain in all three then please restore it to `add-page.css`.

**Review by basil — COMMENTED** [review]
Thanks for the quick reply! This looks mostly ready to merge and interactive testing shows no issues. Once the handful of remaining points are addressed I will be ready to approve this.

File: war/src/main/scss/modules/buttons.scss
Comment (basil) [reply war/src/main/scss/modules/buttons.scss:11]: To clarify, here is the diff of `.jenkins-button` for the old compiled CSS and the new compiled CSS after this PR:

```
@@ -909,6 +867,7 @@
   --item-box-shadow--focus: var(--button-box-shadow--focus);
   position: relative;
   z-index: 0;
+  border-radius: 0.66rem;
   -webkit-appearance: none;
   -moz-appearance: none;
   appearance: none;
```

Here is the new CSS:

```
.jenkins-button {
  --item-background--hover: var(--button-background--hover);
  --item-background--active: var(--button-background--active);
  --item-box-shadow--focus: var(--button-box-shadow--focus);
  position: relative;
  z-index: 0;
  border-radius: 0.66rem;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  border: none;
  outline: none;
  margin: 0;
  padding: 0.5rem 0.85rem;
  font-size: 0.8rem;
  font-weight: 500;
  text-decoration: none !important;
  background: transparent;
  color: var(--text-color) !important;
  border-radius: 0.66rem;
  cursor: pointer;
  min-height: 36px;
  white-space: nowrap;
  gap: 1ch;
}
```

So this is not so much a bug per se but rather the introduction of a duplicate `border-radius` where there was no duplicate before.

Comment (janfaracik) [reply war/src/main/scss/modules/buttons.scss:11]: Ah seems like SCSS isn't removing duplicates like LESS was in that case. Should be okay now.

File: war/src/main/scss/abstracts/theme.scss
Comment (janfaracik) [reply war/src/main/scss/abstracts/theme.scss:25]: Good spot! The 'New Item' page imports 'styles.css' anyways so the additional theme import isn't necessary 👍

File: war/src/main/scss/base/spacing.scss
Comment (janfaracik) [reply war/src/main/scss/base/spacing.scss:29]: Fixed now :)

**Review by basil — APPROVED** [review]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Comment by timja** [comment]
Merging a bit sooner as every renovate PR causes this to conflict

**Comment by aneveux** [comment]
Hello @janfaracik :wave: 

We noticed in some proprietary plugins of ours that some buttons were disapearing on mouse hover (the button was still there, just not visible).

We tracked down the changes and `git bisect` led us to this PR, in particular commit 8b2c0409ed5fb89605addc3c43ffd90762c63533.

It looks like after this PR, the classes:

```
  --btn-primary-bg-active
  --btn-primary-bg-hover 
  --btn-primary-bg-focus 
```

are not defined anymore.

Was that done on purpose, or is it something that was missed during the migration?

It is still happening with the latest Jenkins version, so I would assume it didn't lead to any issues in OSS plugins, but just wanted to confirm that this change was intentional.

Thanks a lot for your help.

**Comment by dwnusbaum** [comment]
To clarify the above, the actual problem seems to be the use of the `lighten` function for `--btn-primary-bg-hover` and `--btn-primary-bg-active` which I think has slightly different syntax in LESS vs SCSS and so the migration results in invalid CSS as far as I understand because it did not adapt the syntax. See https://github.com/jenkinsci/jenkins/pull/8425.

**Comment by basil** [comment]
Failed Stylelint upgrades appear to be pointing to more invalid CSS:

#8238
#8404

**Comment by basil** [comment]
Namely:

```
src/main/scss/abstracts/_theme.scss
 259:21  ✖  Unexpected invalid media query "(max-width: breakpoints.$tablet-breakpoint)"  media-query-no-invalid

src/main/scss/base/_layout-commons.scss
 36:10  ✖  Unexpected invalid media query "(min-width: breakpoints.$tablet-breakpoint)"  media-query-no-invalid
 99:8   ✖  Unexpected invalid media query "(max-width: breakpoints.$tablet-breakpoint)"  media-query-no-invalid

src/main/scss/components/_side-panel-tasks.scss
 13:10  ✖  Unexpected invalid media query "(min-width: breakpoints.$tablet-breakpoint)"  media-query-no-invalid

src/main/scss/pages/_about.scss
 10:3  ✖  Unexpected unknown property "mask-border-source"  property-no-unknown
 11:3  ✖  Unexpected unknown property "mask-border-slice"   property-no-unknown
```

**Pull request merged**
